### DISCUSSION TITLE
Openvpn-admin support for IMDSv2

### DISCUSSION THREAD

**Discussion by marinalimeira** [gruntwork-io/knowledge-base#201] [Help] [answered]
I'm having issues getting credentials using gruntwork terraform-aws-openvpn-admin utility after enabling instance metadata version 2 (IMDSv2) on an ec2 openvpn instance. Is IMDSv2 already supported and what is the possible workaround on this?

```
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/main.go:18 (0x839517)
/usr/local/go/src/runtime/proc.go:201 (0x42bf27)
/usr/local/go/src/runtime/asm_amd64.s:1333 (0x456831)
" error="Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors"

```

**Comment by marinalimeira** [comment]
As per https://github.com/gruntwork-io/terraform-aws-openvpn/pull/144, the `openvpn` module supports IMDSv2.

What command are you running to create the credentials?

**Reply by melvinpambago-mf** [reply]
Hi @marinalimeira thank you for your feedback.
Here is the command I am using below:
` openvpn-admin request --aws-region us-east-2 --username <my-aws-username>`

Is there a specific version that support IMDSv2?
I am using openvpn-admin version `v0.5.4`

**Reply by ina-stoyanova** [reply]
Hey @melvinpambago-mf [this](https://github.com/gruntwork-io/terraform-aws-openvpn/releases/tag/v0.19.0) (v0.19.0) is the earliest release that starts supporting IMDSv2.

**Reply by melvinpambago-mf** [reply]
Hi @ina-stoyanova, 

I updated my `openvpn-admin` utility and still have the same issue.

```bash
#openvpn-admin --version
openvpn-admin version v0.19.0
```

### Here is the server openvpn-admin-request log:
```bash
/var/log# tail -20f openvpn-admin-requests.log 
Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated. 
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors
time="2022-02-17T23:57:58Z" level=error msg="*awserr.baseError Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/vendor/github.com/gruntwork-io/gruntwork-cli/errors/errors.go:37 (0x52b633)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/aws_helpers/iam.go:40 (0x82ad35)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/aws_helpers/sqs.go:109 (0x82bec1)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/aws_helpers/sqs.go:122 (0x82bfb6)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/app/cmd_process_certificate_requests.go:69 (0x8344cd)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/app/cmd_process_certificate_requests.go:36 (0x83414f)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/vendor/github.com/gruntwork-io/gruntwork-cli/errors/errors.go:93 (0x52b9b6)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/vendor/github.com/urfave/cli/app.go:485 (0x519ba8)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/vendor/github.com/urfave/cli/command.go:193 (0x51ad4e)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/vendor/github.com/urfave/cli/app.go:250 (0x517da7)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/vendor/github.com/gruntwork-io/gruntwork-cli/entrypoint/entrypoint.go:21 (0x550ad6)
/go/src/github.com/gruntwork-io/package-openvpn/modules/openvpn-admin/src/main.go:18 (0x839517)
/usr/local/go/src/runtime/proc.go:201 (0x42bf27)
/usr/local/go/src/runtime/asm_amd64.s:1333 (0x456831)
" error="Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): NoCredentialProviders: no valid providers in chain. Deprecated.
        For verbose messaging see aws.Config.CredentialsChainVerboseErrors"
```

### Request Output
```bash
[openvpn-admin] INFO[2022-02-18T07:58:48+08:00] Submitting request for new certificate to https://sqs.us-east-2.amazonaws.com/129732931952/openvpn-response-8bc846d8-904d-11ec-9cd2-66cc4064fa36
[openvpn-admin] INFO[2022-02-18T07:58:48+08:00] Waiting for response from OpenVPN server
ERROR: Failed to receive messages on https://sqs.us-east-2.amazonaws.com/129732931952/openvpn-response-8bc846d8-904d-11ec-9cd2-66cc4064fa36 within 300 seconds
```

**Reply by antmordel** [reply]
Hi @melvinpambago-mf, just trying to understand the issue. From the machine you are running `openvpn-admin` from, do you have the _aws profile_ set up? In my case, I can verify this with any of the following alternatives:
1. I can check if I'm running the command prefixed by aws vault: `aws-vault exec ...`
2. I can ask my terminal to print if I selected an AWS_PROFILE: `echo $AWS_PROFILE`
3. I can ask my terminal to print the secret key: `echo $AWS_SECRET_ACCESS_KEY`
4. But ultimately, the most powerful command is: `aws sts get-caller-identity` should return successfully. 

Please let us know if that helps.

**Reply by melvinpambago-mf** [reply]
Hello @antmordel, I am running openvpn-admin request from my local machine and yes I am also authenticated to the same AWS account I am requesting.

![image](https://user-images.githubusercontent.com/82001841/154683478-59a1c272-002e-4067-9c61-f0ce29d36375.png)

**Reply by melvinpambago-mf** [reply]
Hello @antmordel @ina-stoyanova any suggestions on this issue?
TIA

**Reply by antmordel** [reply]
Hi @melvinpambago-mf - I think you are right and maybe we are hitting issues with IMDSv2. Could you confirm which version of openvpn-server are you running? This can be checked by looking at the packer script you run when you deployed the server (example of the [packer script](https://github.com/gruntwork-io/terraform-aws-openvpn/tree/master/examples/packer)). It should be version [v0.19.0](https://github.com/gruntwork-io/terraform-aws-openvpn/releases/tag/v0.19.0) or greater.

Alternatively, you can install this module into your ec2, with the proper version, using gruntwork-installer: https://github.com/gruntwork-io/terraform-aws-openvpn/tree/master/modules/openvpn-admin.

Thanks a lot.

**Reply by melvinpambago-mf** [reply]
Hi @antmordel - It seems not compatible with terraform version 12.29 and suggesting me to update to at least version 12.31.  I am having the following errors below during terraform plan using [v0.19.1](https://github.com/gruntwork-io/terraform-aws-openvpn/releases/tag/v0.19.1)

```
No provider "aws" plugins meet the constraint ">= 3.66.0,>= 3.66.0,~> 3.0".

The version constraint is derived from the "version" argument within the
provider "aws" block in configuration. Child modules may also apply
provider version constraints. To view the provider versions requested by each
module in the current configuration, run "terraform providers".

To proceed, the version constraints for this provider must be relaxed by
either adjusting or removing the "version" argument in the provider blocks
throughout the configuration.


Warning: registry.terraform.io: This version of Terraform has an outdated GPG key and is unable to verify new provider releases. Please upgrade Terraform to at least 0.12.31 to receive new provider updates. For details see: https://discuss.hashicorp.com/t/hcsec-2021-12-codecov-security-event-and-hashicorp-gpg-key-exposure/23512



Warning: registry.terraform.io: This version of Terraform has an outdated GPG key and is unable to verify new provider releases. Please upgrade Terraform to at least 0.12.31 to receive new provider updates. For details see: https://discuss.hashicorp.com/t/hcsec-2021-12-codecov-security-event-and-hashicorp-gpg-key-exposure/23512



Error: no suitable version is available


ERRO[0115] Hit multiple errors:
Hit multiple errors:
exit status 1
```

I am using also openvpn-server version 0.19.1 on packer script which AMI was successfully built.

"BASH_COMMONS_VERSION=v0.1.8"
"PACKAGE_OPENVPN_VERSION=v0.19.1"
```

      "echo '===========bash commons==========='",
      "gruntwork-install --module-name 'bash-commons' --tag $BASH_COMMONS_VERSION --repo https://github.com/gruntwork-io/bash-commons",
      "echo '===========download-openvpn========'",
      "gruntwork-install --module-name 'install-openvpn' --tag $PACKAGE_OPENVPN_VERSION --repo https://github.com/gruntwork-io/package-openvpn",
      "echo '===========init-openvpn==========='",
      "gruntwork-install --module-name 'init-openvpn' --tag $PACKAGE_OPENVPN_VERSION --repo https://github.com/gruntwork-io/package-openvpn",
      "echo '===========start-openvpn-admin==========='",
      "gruntwork-install --module-name 'start-openvpn-admin' --tag $PACKAGE_OPENVPN_VERSION --repo https://github.com/gruntwork-io/package-openvpn",
      "echo '===========backup-openvpn-pki==========='",
      "gruntwork-install --module-name 'backup-openvpn-pki' --tag $PACKAGE_OPENVPN_VERSION --repo https://github.com/gruntwork-io/package-openvpn",
      "echo '===========openvpn-admin==========='",
      "gruntwork-install --binary-name 'openvpn-admin' --tag $PACKAGE_OPENVPN_VERSION --repo https://github.com/gruntwork-io/package-openvpn",
      "echo '============install-openvpn========'",
      "sudo /usr/local/bin/install-openvpn",



```

**Comment by marinalimeira** [comment]
@melvinpambago-mf We are manually updating our code to not use the AWS provider v4.0. You need to set `version = "< 4.0"` in your provider so it doesn't use v4.0. It's backward incompatible and we haven't updated all modules yet.

```
required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "< 4.0"
    }
  }
  ```

**Reply by melvinpambago-mf** [reply]
@marinalimeira tried to setup it up with "> 4.0" and still has this error. Btw I believe the format you've given for required_providers is terraform 13 and above. Thanks

![image](https://user-images.githubusercontent.com/82001841/156071638-25a35f58-5afd-4bb2-96c8-e3205be4e85e.png)

Here's my main.tf snippet

```
provider "aws" {
  # The AWS region in which all resources will be created
  region = var.aws_region

  # Provider version 4.X series is the latest.
  version = "> 4.0"

  # Only these AWS Account IDs may be operated on by this template
  allowed_account_ids = [var.aws_account_id]
}

# ---------------------------------------------------------------------------------------------------------------------
# CONFIGURE REMOTE STATE STORAGE
# ---------------------------------------------------------------------------------------------------------------------

terraform {
  # The configuration for this backend will be filled in by Terragrunt
  backend "s3" {}

  # Only allow this Terraform version. Note that if you upgrade to a newer version, Terraform won't allow you to use an
  # older version, so when you upgrade, you should upgrade everyone on your team and your CI servers all at once.
  required_version = "= 0.12.29"
}
```

**Reply by antmordel** [reply]
Hi @melvinpambago-mf - Could you try doing `terraform init -upgrade`. Thank you so much!

**Reply by antmordel** [reply]
Hi @melvinpambago-mf - did you solve this issue? Are you having still troubles? Many thanks!

**Reply by melvinpambago-mf** [reply]
Hi @antmordel - I'm still having the same issue after executing the given command. I am thinking if the issue here is - the new openvpn module version `v0.19.0` and above is now using updated provider `3.66.0` for `openvpn.backup_bucket` and not yet compatible with our currently installed terraform version (see attached images below) unless we upgrade it in the future. 

![image](https://user-images.githubusercontent.com/82001841/156671475-116c8303-691a-488b-9435-26aeec45cfef.png)

Let me know your thoughts about it.

Much appreciated and thanks for your feedback.

**Reply by melvinpambago-mf** [reply]
Hi @antmordel - Any thoughts about this if my statement in my previous reply were correct?

**Reply by antmordel** [reply]
Hi @melvinpambago-mf  - So if you are trying to use `openvpn-admin` in the OpenVPN server instance, which is in turn authenticating with [instance metadata against AWS](https://blog.gruntwork.io/authenticating-to-aws-with-instance-metadata-b6d812a86b40), and we are try to enable IMDSv2 but we cannot update the Terraform module, I think we cannot progress further.

Did you find any other root cause? Many thanks!

**Reply by antmordel** [reply]
Hi @melvinpambago-mf - did you solve this issue?

**Reply by melvinpambago-mf** [reply]
Hi @antmordel not yet. We are still on the process of upgrading our code's tf supported version to proceed which will take us bit longer to complete.

**Comment by rhoboat** [comment, marked as answer]
Within the threads, this has been "solved" as an issue that will be resolved once upgrading the openvpn module to at least [v0.17.0](https://github.com/gruntwork-io/terraform-aws-openvpn/releases/tag/v0.17.0). However, in this case they're already using a newer version [v0.19.0](https://github.com/gruntwork-io/terraform-aws-openvpn/releases/tag/v0.19.0).

However, starting in v0.17.1, we actually introduced a backward incompatibility with Terraform < 0.13! Here is the PR that introduced a dependency on [private-s3-bucket v0.57.1](https://github.com/gruntwork-io/terraform-aws-security/blob/v0.57.1/modules/private-s3-bucket/main.tf#L8). As you can see from that line, `required_providers` was already in use by that version (started in [v0.48.3](https://github.com/gruntwork-io/terraform-aws-security/blob/v0.48.3/modules/private-s3-bucket/main.tf#L7)).

That module, private-s3-bucket, requires using Terraform AWS Provider version `>= 3.66.0`, in order to setup the new `aws_s3_replication_configuration`.

My understanding based on all of this, is that we did not adequately communicate that there was a backward incompatibility introduced in every module that was updated to use _any other module_ that uses `required_providers` _and also_ required an AWS Provider version that would cause the module to fail if Terraform < 0.13 was used.

That is, the openvpn repo update to v0.17.1 and above should have mentioned this in the guide.
